File: 2025_Oct/Jayden/2025-11-30-Permutations.java
class Solution {
    /**
        Time Complexity: O(n * n!)
        Space Complexity: O(n * n!)
            - Each permuatation has n integers and there are n! permutations
     */
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(nums, new ArrayList<>(), result);
        return result;
    }

    private void backtrack(int[] nums, List<Integer> path, List<List<Integer>> result) {
        if (path.size() == nums.length) {

            // must create a new list becuase path will get deleted afterwards so result will only contains the empty lists at the end
            result.add(new ArrayList<>(path));
            return;
        }

        // try placing each number in current position
        for (int num : nums) {
            if (path.contains(num)) continue;

            path.add(num);

            // go further with this combination
            backtrack(nums, path, result);

            // backtrack so we can try another number
            path.remove(path.size() - 1);

        }
    }
}

File: 2025_Oct/Kyle/2025-11-30-Permutations.py
class Solution(object):
    def permute(self, nums):
        output = []

        def backtrack(curr):
            if len(curr) == len(nums):
                output.append(curr[:])
                return

            for num in nums:
                if num not in curr:
                    curr.append(num)
                    backtrack(curr)
                    curr.pop()

        backtrack([])

        return output